File: apps/android/app/src/main/java/com/exponential/app/ui/components/EndedRunRow.kt
package com.exponential.app.ui.components

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.exponential.app.ui.icons.ExpIcons
import com.exponential.app.ui.theme.DesignTokens
import com.exponential.app.ui.theme.GlassTokens
import com.exponential.app.ui.theme.TextEmphasis
import com.exponential.app.ui.theme.glassRow

/**
 * EXP-637: one run in a runs list — the Actions screen's "Recent automated
 * runs" and the Agents screen's "Past".
 *
 * EXP-773 made it a plain LINK. A row used to expand to the agent's close-out
 * summary and a Resume pill; both now live at the top of the fullscreen
 * session view, next to the transcript they belong to. So a row is title ·
 * state · byline and a tap opens that session — live or finished, the same
 * gesture, the same destination. Same rule on web, desktop and iOS.
 */
@Composable
fun EndedRunRow(
    title: String,
    timeLabel: String,
    onOpen: () -> Unit,
    modifier: Modifier = Modifier,
    // The monospace lead-in an issue-scoped run shows (its identifier); action
    // and chat runs have none.
    identifier: String? = null,
    // The machine that ran it, when the surface tracks one.
    deviceLabel: String? = null,
    /**
     * EXP-746: the Agents screen's "Past" caption, composed once by
     * `pastRunByline` — `<device> · <time>` (EXP-833).
     * When set it REPLACES the device/time line below, so the ×4 string is
     * whatever that one function produced; the Automations list passes none
     * and keeps the caption it always had.
     */
    byline: String? = null,
    // The run is still going: "Running".
    isLive: Boolean = false,
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .testTag("ended-run-row")
            .glassRow()
            .clickable(onClick = onOpen)
            .padding(horizontal = GlassTokens.RowPaddingH, vertical = GlassTokens.RowPaddingV),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Column(modifier = Modifier.weight(1f)) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    if (identifier != null) {
                        Text(
                            identifier,
                            style = MaterialTheme.typography.labelMedium,
                            fontFamily = FontFamily.Monospace,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
                            maxLines = 1,
                        )
                        Spacer(Modifier.width(8.dp))
                    }
                    Text(
                        title,
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurface,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                }
                Spacer(Modifier.height(2.dp))
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(4.dp),
                ) {
                    if (isLive) {
                        // Live reads green like every other "running now"
                        // signal (iOS parity); a finished run says nothing.
                        Text(
                            "Running",
                            style = MaterialTheme.typography.bodySmall,
                            color = DesignTokens.Semantic.Green,
                            maxLines = 1,
                        )
                    }
                    // "started 5m ago" while it runs, "ended 5m ago" once it
                    // finished — iOS `runByline` parity. EXP-746: a finished
                    // row given a composed byline shows that instead.
                    val trailing = byline?.takeIf { it.isNotBlank() && !isLive }
                        ?: listOfNotNull(
                            deviceLabel?.takeIf { it.isNotBlank() },
                            timeLabel.takeIf { it.isNotEmpty() }
                                ?.let { if (isLive) "started $it" else "ended $it" },
                        ).joinToString(" · ")
                    if (trailing.isNotEmpty()) {
                        Text(
                            // The dot only separates — nothing precedes it on
                            // a finished row anymore.
                            if (isLive) "· $trailing" else trailing,
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis,
                        )
                    }
                }
            }
            Icon(
                ExpIcons.uiChevronRight,
                contentDescription = null,
                modifier = Modifier.padding(start = 8.dp).size(16.dp),
                tint = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
            )
        }
    }
}

File: apps/ios/ExpUI/Sources/EndedRunRow.swift
import SwiftUI

/// EXP-637: the ONE row a runs list draws — the Actions tab's "Recent
/// automated runs" and the Agent page's "Recent" and the issue detail's "Runs" (EXP-886).
///
/// EXP-773 made it a plain LINK. A row used to expand to the agent's close-out
/// summary and a Resume pill; both now live at the top of the fullscreen
/// session view, where the transcript they belong to is. So a row is title ·
/// state · byline and a tap opens that session — live or finished, the same
/// gesture, the same destination. The same rule holds on web, desktop and
/// Android.
///
/// EXP-818: it wears the FLAT list row (`.flatRow()`) — both its lists sit
/// under a `GlassSectionBand` now, and a bordered card inside a table was the
/// one row that still read as a card.
public struct EndedRunRow: View {
    private let title: String
    private let identifier: String?
    private let byline: String
    private let isLive: Bool
    /// EXP-897: this run has child runs nested under it — the row carries the
    /// fold chevron, in a plain Button OUTSIDE the row's own button label.
    private let expandable: Bool
    private let expanded: Bool
    private let onToggle: (() -> Void)?
    private let onOpen: () -> Void

    /// - Parameter isLive: the run is still going — the row says so; the tap
    ///   target is the same either way.
    public init(
        title: String,
        identifier: String? = nil,
        byline: String,
        isLive: Bool = false,
        expandable: Bool = false,
        expanded: Bool = true,
        onToggle: (() -> Void)? = nil,
        onOpen: @escaping () -> Void
    ) {
        self.title = title
        self.identifier = identifier
        self.byline = byline
        self.isLive = isLive
        self.expandable = expandable
        self.expanded = expanded
        self.onToggle = onToggle
        self.onOpen = onOpen
    }

    public var body: some View {
        HStack(alignment: .top, spacing: 6) {
            foldControl
            Button(action: onOpen) {
                header
            }
            .buttonStyle(.plain)
            // The styleguide capture taps this to reach the session view
            // (EXP-663); same tag as Android's.
            .accessibilityIdentifier("ended-run-row")
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 12)
        .flatRow()
    }

    /// EXP-897: the fold. A control inside the row's own Button label would
    /// never receive the tap — it lives beside it.
    @ViewBuilder
    private var foldControl: some View {
        if expandable {
            Button { onToggle?() } label: {
                AppIcon(
                    expanded ? AppIcons.uiChevronDown : AppIcons.uiChevronRight, size: 12
                )
                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                .frame(width: 14, height: 20)
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityLabel(expanded ? "Collapse child runs" : "Expand child runs")
            .accessibilityIdentifier("session-fold")
        }
    }

    private var header: some View {
        HStack(spacing: 10) {
            VStack(alignment: .leading, spacing: 3) {
                HStack(spacing: 6) {
                    if let identifier, !identifier.isEmpty {
                        Text(identifier)
                            .font(.caption.monospaced())
                            .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                            .lineLimit(1)
                    }
                    Text(title)
                        .font(.subheadline.weight(.medium))
                        .foregroundStyle(.white)
                        .lineLimit(1)
                }
                HStack(spacing: 6) {
                    // EXP-686: "Running" is the ONLY status word left — a
                    // finished row just carries its byline.
                    if isLive {
                        Text("Running")
                            .font(.caption.weight(.semibold))
                            .foregroundStyle(DesignTokens.Semantic.green)
                            .lineLimit(1)
                    }
                    if !byline.isEmpty {
                        Text(byline)
                            .font(.caption)
                            .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                            .lineLimit(1)
                    }
                }
            }

            Spacer(minLength: 0)

            AppIcon(AppIcons.uiChevronRight, size: AppIcon.Size.small)
                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
        }
        .contentShape(Rectangle())
    }
}
